src/apprt/gtk/menu.zig

```diff
@@ -0,0 +1,136 @@
+const std = @import("std");
+
+const c = @import("c.zig").c;
+const apprt = @import("../../apprt.zig");
+const App = @import("App.zig");
+const Window = @import("Window.zig");
+const Surface = @import("Surface.zig");
+const Builder = @import("Builder.zig");
+
+const gtk = @import("gtk");
+const gdk = @import("gdk");
+const gio = @import("gio");
+const gobject = @import("gobject");
+
+const log = std.log.scoped(.gtk_menu);
+
+pub fn Menu(
+    comptime T: type,
+    comptime variant: []const u8,
+    comptime style: enum { popover_menu, popover_menu_no_arrow, popover_menu_bar },
+) type {
+    return struct {
+        const Self = @This();
+        const MenuWidget = switch (style) {
+            .popover_menu => gtk.PopoverMenu,
+            .popover_menu_no_arrow => gtk.PopoverMenu,
+            .popover_menu_bar => gtk.PopoverMenuBar,
+        };
+
+        parent: *T,
+        menu_widget: *MenuWidget,
+
+        pub fn init(self: *Self) void {
+            const name = switch (T) {
+                Window => "window",
+                Surface => "surface",
+                else => unreachable,
+            };
+            const parent: *T = @alignCast(@fieldParentPtr(variant, self));
+
+            var builder = Builder.init("menu-" ++ name ++ "-" ++ variant, .ui);
+            defer builder.deinit();
+
+            const object = builder.getObject("menu") orelse unreachable;
+            const menu_model: *gio.MenuModel = @ptrCast(@alignCast(object));
+
+            const menu_widget: *MenuWidget = switch (style) {
+                .popover_menu, .popover_menu_no_arrow => brk: {
+                    const menu_widget: *MenuWidget = MenuWidget.newFromModelFull(menu_model, .{ .nested = true });
+                    if (style == .popover_menu_no_arrow) {
+                        menu_widget.as(gtk.Popover).setHasArrow(0);
+                    }
+                    _ = gtk.Popover.signals.closed.connect(
+                        menu_widget,
+                        *Self,
+                        gtkRefocusTerm,
+                        self,
+                        .{},
+                    );
+                    break :brk menu_widget;
+                },
+                .popover_menu_bar => brk: {
+                    break :brk gtk.PopoverMenuBar.newFromModel(menu_model);
+                },
+            };
+
+            self.* = .{
+                .parent = parent,
+                .menu_widget = menu_widget,
+            };
+        }
+
+        pub fn setParent(self: *const Self, widget: *c.GtkWidget) void {
+            self.menu_widget.as(gtk.Widget).setParent(@ptrCast(@alignCast(widget)));
+        }
+
+        pub fn asWidget(self: *const Self) *c.GtkWidget {
```

Comment:
We should try to avoid c.Gtk* types in new code anyway. Any old code that still expects those types should call @ptrCast themselves

Reply:
Done.